### DISCUSSION TITLE
How to make request body mandatory

### DISCUSSION THREAD

**Discussion by vrjram** [fastapi/fastapi#7802] [Questions] [answered]
Hi,
I have a code segment as below:

```
@app.post('/infer/')
async def infer(request: Request = Body(...)):
```
But in the API spec, the request body is not defined as a mandatory field. Infact there is no body.
```
"/infer/": {
            "post": {
                "summary": "Infer",
                "operationId": "infer_infer__post",
                "responses": {
                    "200": {
                        "description": "Successful Response",
                        "content": {
                            "application/json": {
                                "schema": {}
                            }
                        }
                    }
                }
            }
        }
```

Is there any additional steps required to make this mandatory. Please advice.

Thank you

**Comment by alfrcr** [comment]
> The same as when declaring query parameters, when a model attribute has a default value, it is not required. Otherwise, it is required. Use None to make it just optional.

Taken from https://fastapi.tiangolo.com/tutorial/body/#create-your-data-model
Hope this helps

**Comment by nsidnev** [comment]
Also, if you just want your request to require some kind of json body without defining a pydantic model, you can define it like this:

```python3
from fastapi import FastAPI

app = FastAPI()


@app.post("/my-route")
async def my_post_route(request_data: dict) -> dict:
    return {"posted": request_data}
```

**Comment by pawamoy** [comment]
I thought using an ellipsis would make the body parameter mandatory as well, just like query parameters: https://fastapi.tiangolo.com/tutorial/query-params-str-validations/#make-it-required. If not, I would consider this a bug, or at least inconsistent behavior 🙂 

Maybe it does not work as expected in this particular case because your `request` parameter has `Request` for type, which if I guess correctly is imported from `starlette.requests`? I'm not even sure what it would mean semantically: a body parameter which is also a request 😕 ? Or maybe you defined a `Request` pydantic model?

**Comment by vrjram** [comment, marked as answer]
> I thought using an ellipsis would make the body parameter mandatory as well, just like query parameters: https://fastapi.tiangolo.com/tutorial/query-params-str-validations/#make-it-required. If not, I would consider this a bug, or at least inconsistent behavior 🙂
> 
> Maybe it does not work as expected in this particular case because your `request` parameter has `Request` for type, which if I guess correctly is imported from `starlette.requests`? I'm not even sure what it would mean semantically: a body parameter which is also a request 😕 ? Or maybe you defined a `Request` pydantic model?

I understand that it doesnt make sense as I am using starlette Requests here. Since then the code has been refactored to use a pydantic model. Thank you

**Comment by vrjram** [comment]
> Also, if you just want your request to require some kind of json body without defining a pydantic model, you can define it like this:
> 
> ```python
> from fastapi import FastAPI
> 
> app = FastAPI()
> 
> 
> @app.post("/my-route")
> async def my_post_route(request_data: dict) -> dict:
>     return {"posted": request_data}
> ```

Thank you for the response. This was required for a quick POC where I couldnt finalize the pydantic model initially.

**Comment by tiangolo** [comment]
Thanks for the help here everyone! :cake: 

And thanks @vrjram for reporting back and closing the issue :+1:
